compiler/rustc_macros/src/diagnostics/diagnostic.rs

```diff
@@ -71,31 +71,46 @@ impl<'a> SessionDiagnosticDerive<'a> {
                     }
                 };
 
+                // Keep track of which fields are subdiagnostics
+                let mut subdiagnostics = std::collections::HashSet::new();
+
                 // Generates calls to `span_label` and similar functions based on the attributes
                 // on fields. Code for suggestions uses formatting machinery and the value of
                 // other fields - because any given field can be referenced multiple times, it
                 // should be accessed through a borrow. When passing fields to `set_arg` (which
                 // happens below) for Fluent, we want to move the data, so that has to happen
                 // in a separate pass over the fields.
-                let attrs = structure.each(|field_binding| {
-                    let field = field_binding.ast();
-                    let result = field.attrs.iter().map(|attr| {
-                        builder
-                            .generate_field_attr_code(
-                                attr,
-                                FieldInfo {
-                                    vis: &field.vis,
-                                    binding: field_binding,
-                                    ty: &field.ty,
-                                    span: &field.span(),
-                                },
-                            )
-                            .unwrap_or_else(|v| v.to_compile_error())
+                let attrs = structure
+                    .clone()
+                    // Remove the fields that have a `subdiagnostic` attribute.
+                    .filter(|field_binding| {
```

Comment:
I think it might be better if we do the following:

- Add the `set_arg` generation to `generate_field_attr_code` so that in both of the `structure.each(|| { .. })` closures we can just call that function.
- Create a filter function which checks for no attributes or a `subdiagnostic` attribute and use that for both `structure.each(|| { .. })` calls (negated the first time, not the second).

I think this could be a little bit tidier.